pkg/http/auth_middleware.go

```diff
@@ -0,0 +1,339 @@
+	hash := hmac.New(sha256.New, key)
+	hash.Write(data)
+
+	return hash.Sum(nil)
+}
+
+func getHashedPayload(req http.Request) string {
+	hashedPayload := req.Header.Get("X-Amz-Content-Sha256")
+	if hashedPayload == "" {
+		// Presign does not have a payload, use S3 recommended value.
+		hashedPayload = unsignedPayload
+	}
+
+	return hashedPayload
+}
+
+func headerExists(key string, headers []string) bool {
+	for _, k := range headers {
+		if k == key {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getHostAddr returns host header if available, otherwise returns host from URL
+func getHostAddr(req *http.Request) string {
+	host := req.Header.Get("host")
+	if host != "" && req.Host != host {
+		return host
+	}
+
+	if req.Host != "" {
+		return req.Host
+	}
+
+	return req.URL.Host
+}
+
+// Trim leading and trailing spaces and replace sequential spaces with one space, following Trimall()
+// in http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+func signV4TrimAll(input string) string {
+	// Compress adjacent spaces (a space is determined by
+	// unicode.IsSpace() internally here) to one space and return
+	return strings.Join(strings.Fields(input), " ")
+}
+
+func getCanonicalHeaders(req http.Request, signedHeaders string) string {
+	headers := []string{}
+	vals := make(map[string][]string)
+
+	for _, k := range strings.Split(signedHeaders, ";") {
+		vv := req.Header.Values(k)
+		headers = append(headers, strings.ToLower(k))
+		vals[strings.ToLower(k)] = vv
+	}
+
+	if !headerExists("host", headers) {
+		headers = append(headers, "host")
+	}
+
+	sort.Strings(headers)
+
+	var buf bytes.Buffer
+	// Save all the headers in canonical form <header>:<value> newline
+	// separated for each header.
+	for _, k := range headers {
+		buf.WriteString(k)
+		buf.WriteByte(':')
+
+		switch {
+		case k == "host":
+			buf.WriteString(getHostAddr(&req))
+			buf.WriteByte('\n')
+		default:
+			for idx, v := range vals[k] {
+				if idx > 0 {
+					buf.WriteByte(',')
+				}
+
+				buf.WriteString(signV4TrimAll(v))
+			}
+
+			buf.WriteByte('\n')
+		}
+	}
+
+	return buf.String()
+}
+
+func encodePath(pathName string) string {
+	if reservedObjectNames.MatchString(pathName) {
+		return pathName
+	}
+
+	var encodedPathname strings.Builder
+
+	for _, s := range pathName {
+		if 'A' <= s && s <= 'Z' || 'a' <= s && s <= 'z' || '0' <= s && s <= '9' { // §2.3 Unreserved characters (mark)
+			encodedPathname.WriteRune(s)
+			continue
+		}
+
+		switch s {
+		case '-', '_', '.', '~', '/': // §2.3 Unreserved characters (mark)
+			encodedPathname.WriteRune(s)
+			continue
+		default:
+			sLen := utf8.RuneLen(s)
+			if sLen < 0 {
+				// if utf8 cannot convert return the same string as is
+				return pathName
+			}
+
+			u := make([]byte, sLen)
+			utf8.EncodeRune(u, s)
+
+			for _, r := range u {
+				encodedPathname.WriteString("%" + strings.ToUpper(hex.EncodeToString([]byte{r})))
+			}
+		}
+	}
+
+	return encodedPathname.String()
+}
```

Comment:
Can we use [url.PathEscape](https://github.com/golang/go/blob/master/src/net/url/url.go#L281) instead?

Reply:
I'm not so sure about this as it is taken from what the MinIO client used for encoding the path. I used the same code for the compatibility.

Reply:
As this is [publicly exposed upstream](https://github.com/minio/minio-go/blob/master/pkg/s3utils/utils.go#L327), I think we could just add it as a dependency. The upstream implementation provides context, and given the larger use base would help with long term maintainability/improvements.

Reply:
ha. I should have done that.
v7 of the cli had been imported as a dep already.

thank you for spotting that.